### PR TITLE
Add Two Factor Authentication to Account 

### PR DESCRIPTION
**Pull request by jianba**
_urlEncoder.Encode() cant be used that may cause  garbled

### CODE REVIEW

File: Account/Aiursoft.Account.csproj [hunk 1]
```diff
@@ -16,6 +16,11 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.9.4">
```
Comment (Anduin2017) [inline Account/Aiursoft.Account.csproj]: Seems useless to the project. Do not commit this.

**Comment by Anduin2017** [comment]
For this feature, I will not help you implement it. You have enough time to think and read the existing code. Like email management implementation and phone management implementation is the feature which you can reference. Please leave this pull request open and keep fixing the code.

**Comment by jianba** [comment]
> For this feature, I am not planning to help you with implementing it. You have enough time to think and read the existing code. Like email management feature and the phone management feature are which you can reference. Please leave this pull request open and keep fixing the code.

OK. I will refer to the existing code to correct the current code.

**Comment by Anduin2017** [comment]
BTW. It seems you are struggling with git. You can just leave this pull request open and keep changing (commit to) your master branch. And the pull request will be auto-updated.

**Comment by Anduin2017** [comment]
Hi, while this feature is great, but please clean up your pull request and remove those useless changes. Make the change minimum to achieve the feature is required. Always remember: Keep Nexus code clean for it contains the very core key features for Aiursoft.

File: Account/Controllers/AccountController.cs [hunk 2]
```diff
@@ -31,6 +34,11 @@ public class AccountController : Controller
         private readonly DeveloperApiService _developerApiService;
         private readonly AuthService<AccountUser> _authService;
 
+        private readonly UrlEncoder _urlEncoder;
+        private const string RecoveryCodesKey = nameof(RecoveryCodesKey);
+        private readonly ILogger _logger;
+        private const string AuthenticatorUriFormat = "otpauth://totp/{0}:{1}?secret={2}&issuer={0}&digits=6";
```
Comment (Anduin2017) [inline Account/Controllers/AccountController.cs]: Rename to `_AuthenticatorUriFormat`

File: Account/Controllers/AccountController.cs [hunk 3]
```diff
@@ -31,6 +34,11 @@ public class AccountController : Controller
         private readonly DeveloperApiService _developerApiService;
```
Comment (Anduin2017) [inline Account/Controllers/AccountController.cs]: Sort those fields.

File: Account/Controllers/AccountController.cs (hunk 2 above)
Comment (Anduin2017) [inline Account/Controllers/AccountController.cs]: Rename it to starts with underline.

File: Account/Controllers/AccountController.cs [hunk 4]
```diff
@@ -48,14 +56,15 @@ public class AccountController : Controller
             _developerApiService = developerApiSerivce;
             _authService = authService;
         }
-
+        [HttpGet]
```
Comment (Anduin2017) [inline Account/Controllers/AccountController.cs]: New line before required.

File: Account/Controllers/AccountController.cs [hunk 5]
```diff
@@ -48,14 +56,15 @@ public class AccountController : Controller
...
         public async Task<IActionResult> Index(bool? justHaveUpdated)
         {
             var user = await GetCurrentUserAsync();
             var model = new IndexViewModel(user)
             {
                 JustHaveUpdated = justHaveUpdated ?? false
             };
+
```
Comment (Anduin2017) [inline Account/Controllers/AccountController.cs]: Useless line.

File: Account/Controllers/AccountController.cs [hunk 6]
```diff
@@ -72,7 +81,8 @@ public async Task<IActionResult> Index(IndexViewModel model)
             }
             cuser.NickName = model.NickName;
             cuser.Bio = model.Bio;
-            await _userService.ChangeProfileAsync(cuser.Id, await _appsContainer.AccessToken(), cuser.NickName, cuser.IconFilePath, cuser.Bio);
+            cuser.Bio2 = model.Bio;
+            await _userService.ChangeProfileAsync(cuser.Id, await _appsContainer.AccessToken(), cuser.NickName, cuser.IconFilePath, cuser.Bio ,cuser.Bio2);
```
Comment (Anduin2017) [inline Account/Controllers/AccountController.cs]: Useless Bio2. Remove it.

File: Account/Controllers/AccountController.cs [hunk 7]
```diff
@@ -192,7 +202,7 @@ public async Task<IActionResult> Avatar(AvatarViewModel model)
                 return View(model);
             }
             cuser.IconFilePath = model.NewIconAddres;
-            await _userService.ChangeProfileAsync(cuser.Id, await _appsContainer.AccessToken(), cuser.NickName, cuser.IconFilePath, cuser.Bio);
+            await _userService.ChangeProfileAsync(cuser.Id, await _appsContainer.AccessToken(), cuser.NickName, cuser.IconFilePath, cuser.Bio, cuser.Bio2);
```
Comment (Anduin2017) [inline Account/Controllers/AccountController.cs]: Useless Bio2.

File: Account/Controllers/AccountController.cs [hunk 8]
```diff
@@ -344,6 +354,186 @@ async Task AddApp()
             return View(model);
         }
 
+        [HttpGet]
+        public async Task<IActionResult> TwoFactorAuthentication()
+        {
```
Comment (Anduin2017) [inline Account/Controllers/AccountController.cs]: Why didn't you call API to get current user authentication status? Is authentication key storage in API DB? How can you maintain it without getting info from API?

File: Account/Migrations/20190927122029_addBio2.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
```
Comment (Anduin2017) [inline Account/Migrations/20190927122029_addBio2.cs]: Useless migration.

**Comment by jianba** [comment]
Because I created a new branch, I couldn't continue using this pull request.

**Pull request closed without merging**